ta/pkcs11/src/attributes.c

```diff
@@ -207,6 +207,48 @@ bool get_bool(struct obj_attrs *head, uint32_t attribute)
 	return bbool;
 }
 
+bool attributes_match_reference(struct obj_attrs *candidate,
+				struct obj_attrs *ref)
+{
+	size_t count = ref->attrs_count;
+	unsigned char *ref_attr = ref->attrs;
+	uint32_t rc = PKCS11_CKR_GENERAL_ERROR;
+
+	if (!ref->attrs_count) {
+		DMSG("Empty reference: no match");
+		return false;
+	}
+
+	for (count = 0; count < ref->attrs_count; count++) {
+		struct pkcs11_attribute_head pkcs11_ref = { };
+		void *found = NULL;
+		uint32_t size = 0;
+
+		TEE_MemMove(&pkcs11_ref, ref_attr, sizeof(pkcs11_ref));
+
+		if (pkcs11_is_attr_bool(pkcs11_ref.id)) {
+			bool bb_ref = get_bool(ref, pkcs11_ref.id);
+			bool bb_candidate = get_bool(candidate, pkcs11_ref.id);
+
+			if (bb_ref != bb_candidate)
+				return false;
+		} else {
+			rc = get_attribute_ptr(candidate, pkcs11_ref.id,
+					       &found, &size);
+
+			if (rc || !found || size != pkcs11_ref.size ||
+			    TEE_MemCompare(ref_attr + sizeof(pkcs11_ref),
```

Comment:
This makes an assumption about internal storage in memory.

Should we just get data pointer with get_attribute_ptr(ref, ...) and then compare actual attribute value? That should work in all cases should it be refactored at some point of time.

Reply:
I believe this is OK -- we anyway use size to advance forward. No change needed.